*: un-singletonize the cluster settings

tbg opened the pull request:
Apologies to drop such a huge refactoring, but it was a bit of a coup de force
required (or at least strongly suggested) by my inflight PR on version
migration (#17411).

We were previously using our singleton-based `settings` package that originally
replaced the use of environment variables. Its use, however, has since
diversified, the cluster version setting being the most elaborate addition.

In #17411 it has turned out that with these singletons in place, a lot of
effort and hacks are required to even be able to run tests at the same version.
However, it would be nice to test nonidentical versions in cluster tests.

Consequently, this PR

- introduces a struct `cluster.Settings` which now holds all the settings
  previously scattered around the code base; and
- plumbs this struct everywhere it's needed.

There are some open FIXMEs but I'd like a reviewer to review the bulk that's
here in the initial commit (which should pass all tests). I will then address
the FIXMEs (which require more focused attention) in follow-up commits, before
merging this PR.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17486)
<!-- Reviewable:end -->

RaduBerinde commented:
Awesome!! So happy to get rid of these globals! Thanks for putting in all this work! 🎉 💯

RaduBerinde commented:
It's unfortunate that we have to move the definition of the settings away from the code that uses them.. Can't the registry stay global (de-globalize just the values)? We could still have global `RegisterXXX` calls which associate settings to "slot" indexes; a separate "settings instance" structure would have values for all "slots" and would need to be passed to all get/set calls.

tschottdorf commented:
@RaduBerinde I don't feel strongly one way or another, but I do feel strongly that the plumbing I've done here is all the plumbing I'm going to do in this PR. Once we've gotten rid of the singletons in the first place though, happy to discuss ergonomics!

RaduBerinde commented:
Fair enough! :)

petermattis commented:
I'm :thumbsup: on getting rid of the singleton and :thumbsdown: on centralizing the settings as is being done in `cluster/settings.go`. I like @raduberinde's suggestion which would eliminate a ton of the code changes here. Seems like it would revert a bunch of the changes in this PR, leaving the plumbing of the "settings instance" (i.e. the removal of the singleton settings storage).

---

Review status: 0 of 122 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-:-Kqy-0EE5utNTRM4TMmY:b-g65lce)*
<!-- Sent from Reviewable.io -->

tbg committed:
*: un-singletonize the cluster settings

Apologies to drop such a huge refactoring, but it was a bit of a coup de force
required (or at least strongly suggested) by my inflight PR on version
migration (#17411).

We were previously using our singleton-based `settings` package that originally
replaced the use of environment variables. Its use, however, has since
diversified, the cluster version setting being the most elaborate addition.

In #17411 it has turned out that with these singletons in place, a lot of
effort and hacks are required to even be able to run tests at the same version.
However, it would be nice to test nonidentical versions in cluster tests.

Consequently, this PR

- introduces a struct `cluster.Settings` which now holds all the settings
  previously scattered around the code base; and
- plumbs this struct everywhere it's needed.

There are some open FIXMEs but I'd like a reviewer to review the bulk that's
here in the initial commit (which should pass all tests). I will then address
the FIXMEs (which require more focused attention) in follow-up commits, before
merging this PR.

nvanbenschoten commented:
Reviewed 125 of 125 files at r1, 18 of 18 files at r2.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/server/config_test.go, line 31 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqyJj7NbEA0F4Y-1l47:-KqyJj7NbEA0F4Y-1l48:b-exf04b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/server/config_test.go#L31)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/util/envutil"
> 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
> 	"github.com/kr/pretty"
> ```

move up

---

*[pkg/settings/bool.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy86DP_Ki8vNDx_-S1:-Kqy86DP_Ki8vNDx_-S2:bzczg1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/settings/bool.go#L53)):*
> ```Go
> 
> func (b *BoolSetting) set(v bool) {
> 	vInt := int32(0)
> ```

The flow of control between `Override` and `set` is different between `IntSetting` and `BoolSetting`. I'd expect they'd act the same way in terms of who calls `changed`, etc. The easiest way to fix this is probably just to have `BoolSetting.set` call `BoolSetting.Override`.

---

*[pkg/settings/registry.go, line 38 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy8Zaubo6xbTBPu671:-Kqy8Zaubo6xbTBPu672:be7jmyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dda451b3b973c56f22717eca76fb6ba45c5b520/pkg/settings/registry.go#L38)):*
> ```Go
> var frozen int32
> 
> // Freeze ensures that no new settings can be defined after the gossip worker
> ```

We're ok abandoning this?

---

*[pkg/settings/cluster/settings.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqxyEleQGYtm60J8zhb:-KqxyEleQGYtm60J8zhc:by54l7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/settings/cluster/settings.go#L46)):*
> ```Go
> 
> const (
> 	// DebugRemoteOff disallows access to /debug/requests.
> ```

Want to typedef this like we do for `DistSQLExecMode`?

---

*[pkg/settings/cluster/settings.go, line 84 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqxyZIEhob7d6K8O9b1:-KqxyZIEhob7d6K8O9b2:b-o29ed3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/settings/cluster/settings.go#L84)):*
> ```Go
> }
> 
> // DistSQLExecModeFromInt converts an int64 into a DistSQLExecMode
> ```

What's the point of this function? It's just a cast.

---

*[pkg/settings/cluster/settings.go, line 136 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqxzfBjQUyDNEzZFQ6-:-KqxzfBjQUyDNEzZFQ60:b-hyu3ub) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/settings/cluster/settings.go#L136)):*
> ```Go
> }
> 
> // SQLSettings is the subset of ClusterSettings affecting SQL.
> ```

These are all related to SQL statistics collection. Calling it `SQLSettings` instead of `SQLStatsSettings`, when `SQLSessionSettings` exists, seems too broad.

---

*[pkg/settings/cluster/settings.go, line 222 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy-X53QAAHa32JS5Pi:-Kqy-X53QAAHa32JS5Pj:byeylkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/settings/cluster/settings.go#L222)):*
> ```Go
> 	DiagnosticReportFrequency *settings.DurationSetting
> 	TimeUntilStoreDead        *settings.DurationSetting
> 	MaxIntents                *settings.IntSetting
> ```

This doesn't belong in `StorageSettings`? Neither does the setting above? I like the pattern of embedded categorized structs. It would be a shame if not all settings fit in somewhere.

---

*[pkg/sql/app_stats.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy9UCDLKaV4d346Btf:-Kqy9UCDLKaV4d346Btg:b-i3l8b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/sql/app_stats.go#L139)):*
> ```Go
> // each node. It hangs off Executor.
> type sqlStats struct {
> 	st cluster.Settings
> ```

We're not passing around a pointer to `cluster.Settings`? Isn't it a huge struct?

---

*[pkg/sql/executor.go, line 250 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy9z6CksGpBzaFGq-X:-Kqy9z6CksGpBzaFGq-Y:b-6x585x) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/sql/executor.go#L250)):*
> ```Go
> // a Executor; the rest will have sane defaults set if omitted.
> type ExecutorConfig struct {
> 	cluster.Settings
> ```

Embedding this struct everywhere seems like a mistake. We're leaking a lot of unnecessary fields up into these structs.

---

*[pkg/sql/logictest/testdata/logic_test/show_source, line 88 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqyBKyF_KYhZ68GgvXO:-KqyBKyGyDGFtViccQwJ:b-iwcwbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/sql/logictest/testdata/logic_test/show_source#L88)):*
> in the format '\<major\>.\<minor\>'

what's the `-2` then? This should mention the patch, right?

Or is that the unstable part of the version? If so, will this be changing a lot and causing a lot of churn on this line?

---

*[pkg/util/log/structured.go, line 151 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqyGqktNo1iSFQvBPcO:-KqyGqktNo1iSFQvBPcP:b5midh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/util/log/structured.go#L151)):*
> ```Go
> 		}
> 		reportable = fmt.Sprintf("%s:%d %s", filepath.Base(file), line, reportable)
> 		sendCrashReport(ctx, nil /* no ReportingSettings available */, reportable, depth+1)
> ```

:frowning_face: 

---

*[pkg/util/tracing/tracer.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqyGzn7ggYIBQIolYJC:-KqyGzn7ggYIBQIolYJD:b70ltqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/util/tracing/tracer.go#L87)):*
> ```Go
> 
> 	// True if tracing to the debug/requests endpoint. Accessed via t.useNetTrace().
> 	_useNetTrace int32 // updated atomically
> ```

What's up with the underscore?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Thanks Nathan! Addressed tracing FIXME + fallout, PTAL.

---

Reviewed 1 of 125 files at r1.
Review status: 95 of 145 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/server/config_test.go, line 31 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqyJj7NbEA0F4Y-1l47:-KqzHBF4X_-WKNpUmGQP:bzhli4z) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/server/config_test.go#L31)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

move up
</blockquote></details>

Like this? I like to leave the arrangement here to the linters.

---

*[pkg/settings/bool.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy86DP_Ki8vNDx_-S1:-KqzHenkGIQWqLHZZk40:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/settings/bool.go#L53)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The flow of control between `Override` and `set` is different between `IntSetting` and `BoolSetting`. I'd expect they'd act the same way in terms of who calls `changed`, etc. The easiest way to fix this is probably just to have `BoolSetting.set` call `BoolSetting.Override`.
</blockquote></details>

Done.

---

*[pkg/settings/registry.go, line 38 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy8Zaubo6xbTBPu671:-KqzHhrRqkaElAHbT-tm:b-r24jfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dda451b3b973c56f22717eca76fb6ba45c5b520/pkg/settings/registry.go#L38)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We're ok abandoning this?
</blockquote></details>

Afaict there's really no temptation at this point to mutate a registry later any more.

---

*[pkg/settings/cluster/settings.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqxyEleQGYtm60J8zhb:-KqzMQOGcXx6diPDYANL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/settings/cluster/settings.go#L46)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Want to typedef this like we do for `DistSQLExecMode`?
</blockquote></details>

Done.

---

*[pkg/settings/cluster/settings.go, line 84 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqxyZIEhob7d6K8O9b1:-KqzLonTEQDOe-MVg4in:bmcqqap) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/settings/cluster/settings.go#L84)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's the point of this function? It's just a cast.
</blockquote></details>

Not my code, but can't help but agree -- removed.

---

*[pkg/settings/cluster/settings.go, line 136 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqxzfBjQUyDNEzZFQ6-:-KqzNIBEX5lJnJxgfjbk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/settings/cluster/settings.go#L136)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

These are all related to SQL statistics collection. Calling it `SQLSettings` instead of `SQLStatsSettings`, when `SQLSessionSettings` exists, seems too broad.
</blockquote></details>

Done.

---

*[pkg/settings/cluster/settings.go, line 222 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy-X53QAAHa32JS5Pi:-KqzNMUar3Vt97fCTr8V:b-v67u5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/settings/cluster/settings.go#L222)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This doesn't belong in `StorageSettings`? Neither does the setting above? I like the pattern of embedded categorized structs. It would be a shame if not all settings fit in somewhere.
</blockquote></details>

Played some golf, see how you like that.

---

*[pkg/sql/app_stats.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy9UCDLKaV4d346Btf:-KqzP4LmPPi8SdU4oBES:b-dd3226) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/sql/app_stats.go#L139)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We're not passing around a pointer to `cluster.Settings`? Isn't it a huge struct?
</blockquote></details>

Yeah, you might be right that it's not a great idea, though I thought it's not sitting on anything we fabricate en masse. Could be wrong though. I'll make this an embedded pointer instead (tomorrow).

---

*[pkg/sql/executor.go, line 250 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy9z6CksGpBzaFGq-X:-KqzPE1rgnRZK_zQT2Os:b4i6f71) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/sql/executor.go#L250)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Embedding this struct everywhere seems like a mistake. We're leaking a lot of unnecessary fields up into these structs.
</blockquote></details>

Agreed, will change, see other comment.

---

*[pkg/sql/logictest/testdata/logic_test/show_source, line 88 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqyBKyF_KYhZ68GgvXO:-KqzOanIIvKjlvKneZ2e:brby9w3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/sql/logictest/testdata/logic_test/show_source#L88)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> in the format '\<major\>.\<minor\>'

what's the `-2` then? This should mention the patch, right?

Or is that the unstable part of the version? If so, will this be changing a lot and causing a lot of churn on this line?
</blockquote></details>

See the versioning RFC (or perhaps better, my open PR on cluster versioning) -- we have the patch version but ignore it (i.e. it's always zero). Instead we use the "unstable" to sequence all the individual migrations that occur in the development cycle. A release version will always be `-0` (which is omitted).
Perhaps we'll want to include the patch version, but either way this isn't the PR in which that is decided.

---

*[pkg/util/log/structured.go, line 151 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqyGqktNo1iSFQvBPcO:-KqzOvQIVZF17_Knl60i:bb77ucv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/util/log/structured.go#L151)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

:frowning_face: 
</blockquote></details>

Yep.

---

*[pkg/util/tracing/tracer.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-KqyGzn7ggYIBQIolYJC:-KqzP-WNhhICV5FfDToW:b-4sk1mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/util/tracing/tracer.go#L87)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's up with the underscore?
</blockquote></details>

There's an accessor that is to be used instead of using this variable directly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-:-KqzPHJ9JK56jTEwPXiM:b-hx3wvk)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 125 files at r1, 53 of 53 files at r5, 13 of 13 files at r6, 39 of 39 files at r7, 4 of 4 files at r8, 1 of 1 files at r9, 6 of 6 files at r10, 34 of 34 files at r11.
Review status: 119 of 145 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/kv/txn_coord_sender_test.go, line 1332 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kr1CTVS9xYNGCmn14gW:-Kr1CTVS9xYNGCmn14gX:b-ssfk2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa3f55a189e39c0dc4710c1894dc7575d68ce1af/pkg/kv/txn_coord_sender_test.go#L1332)):*
> ```Go
> 	db := client.NewDB(sender, s.Clock)
> 
> 	log.Warningf(context.TODO(), "CLOCK IS NOW %s", s.Clock.Now())
> ```

Is this garbage?

---

*[pkg/settings/cluster/settings.go, line 222 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy-X53QAAHa32JS5Pi:-Kr1E7b30kD5EcMxtLdL:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/settings/cluster/settings.go#L222)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Played some golf, see how you like that.
</blockquote></details>

:+1: 

---

*[pkg/sql/app_stats.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy9UCDLKaV4d346Btf:-Kr1CkXAitNt4ArcCIGD:bs1wnnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/sql/app_stats.go#L139)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yeah, you might be right that it's not a great idea, though I thought it's not sitting on anything we fabricate en masse. Could be wrong though. I'll make this an embedded pointer instead (tomorrow).
</blockquote></details>

I think you can also do something like:
```
type SettingsVal struct {
...
}

type Settings *SettingsVal
```

Also, an embedded pointer? You mean a non-embedded pointer like you said in your other comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 119 of 145 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/kv/txn_coord_sender_test.go, line 1332 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kr1CTVS9xYNGCmn14gW:-Kr1FBrmjWCy_DxRZOJr:b-l7u4q1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa3f55a189e39c0dc4710c1894dc7575d68ce1af/pkg/kv/txn_coord_sender_test.go#L1332)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this garbage?
</blockquote></details>

Definitely doesn't get more garbage than this. I was debugging a failure here. Removed!

---

*[pkg/sql/app_stats.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy9UCDLKaV4d346Btf:-Kr1FPRxGjiXuzdxjVkR:be2a8wt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/sql/app_stats.go#L139)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think you can also do something like:
```
type SettingsVal struct {
...
}

type Settings *SettingsVal
```

Also, an embedded pointer? You mean a non-embedded pointer like you said in your other comment?
</blockquote></details>

I'm now embedding a `*cluster.Settings` in most places (that's what I mean with an "embedded pointer"). Were you thinking something else?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 34 files at r11, 1 of 1 files at r12, 25 of 25 files at r13, 1 of 1 files at r14.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/app_stats.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy9UCDLKaV4d346Btf:-Kr1Tc0rBaAn52q1MalN:bv1cw84) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/sql/app_stats.go#L139)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I'm now embedding a `*cluster.Settings` in most places (that's what I mean with an "embedded pointer"). Were you thinking something else?
</blockquote></details>

I had the other comment about not embedding so that the settings fields don't leak up into other structs. Instead, I was thinking `st *cluster.Settings`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486)*
<!-- Sent from Reviewable.io -->

dt commented:
I'm also :thumbsdown: on defining all the settings statically in the registry -- when I've seen this pattern in the past, it's always been regretted, since effectively all packages depend on it and thus any new feature or knob invalidates that package and thus the world. This isn't a huge deal while our codebase is pretty small, but could bite us later. It is particularly an issue with an eye towards plugin/extension support, where those may want to register settings without statically editing the settings package. Thus I too would prefer to plumb the registry (much as it is here), but not introduce static fields and instead stick to init/constructor time registration in said registry.

dt commented:
I should also mention that we can even allow registration of new settings at any time, post init, as long as we then re-read the gossip data. We'd want a bool for if the first read has happened and if it has, rather than panic, we can just queue a re-read to pick up the potential new values.

tschottdorf commented:
I chatted briefly with @RaduBerinde and I think the idea is workable and it also looks that it's the general preference, so in principle I'm for it.

Yet, I realistically don't want to do it in this PR. I'm currently polishing what I have up in the hope of merging it ASAP, once @nvanbenschoten is happy. Churning the settings back to where they originally came from in follow-up work is mildly regrettable, but it's likely more pleasant to pick this up when there's a little less time pressure and the plumbing in this PR is already in place anyway.

If either of @petermattis @dt or @RaduBerinde feel strongly that I shouldn't go ahead with this change as-is, please let me know ASAP.

---

Review status: 129 of 145 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/app_stats.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kqy9UCDLKaV4d346Btf:-Kr1hrBuPV_SjSN8ebjz:b-orvv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccea286862179c46751d0b4a18076b0f024c8eb/pkg/sql/app_stats.go#L139)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I had the other comment about not embedding so that the settings fields don't leak up into other structs. Instead, I was thinking `st *cluster.Settings`.
</blockquote></details>

Done. (`Settings *cluster.Settings` since it usually needs to be accessed publicly).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-:-Kr1f_C-WZByPePW968t:bwiyur7)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 2 of 2 files at r15, 13 of 15 files at r16.
Review status: 143 of 145 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/server/node_test.go, line 273 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kr1ijAre-nR7-3MYtRH:-Kr1ijAs50lbvPMqD9mq:b4mtqot) ([raw file](https://github.com/cockroachdb/cockroach/blob/a73617b8b8de94a153c91603effdeaf4e00ef95d/pkg/server/node_test.go#L273)):*
> ```Go
> }
> 
> // TODO(tschottdorf): tests calling this previously used an empty store config
> ```

What's the TODO? This just seems like a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I don't mind a follow-up PR which moves the setting registration back to globals, but I wouldn't want to leave that lingering too long. Definitely file an issue about doing so. 

---

Review status: 143 of 145 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-:-Kr1jof5Vpm_APVsuB1u:b-tor9bo)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 143 of 145 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/server/node_test.go, line 273 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kr1ijAre-nR7-3MYtRH:-Kr1jDZb_diIGGCNic20:bvy7pzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a73617b8b8de94a153c91603effdeaf4e00ef95d/pkg/server/node_test.go#L273)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's the TODO? This just seems like a comment.
</blockquote></details>

Added "Investigate why." :smile: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486)*
<!-- Sent from Reviewable.io -->

dt commented on pkg/util/log/structured.go:
```diff
@@ -148,7 +148,7 @@ func addStructured(ctx context.Context, s Severity, depth int, format string, ar
 		}
 		reportable = fmt.Sprintf("%s:%d %s", filepath.Base(file), line, reportable)
-		sendCrashReport(ctx, reportable, depth+1)
+		sendCrashReport(ctx, nil /* no ReportingSettings available */, reportable, depth+1)
```
I'm not thrilled about this. Since we can't plumb this one, I'd rather see a log.SendCrashReports bool that gets updated using an OnChange callback in the setting in the registry.

tschottdorf replied:
`sendCrashReport` has a TODO about exactly this which I'll file an issue about (+address it right away).

tschottdorf replied:
https://github.com/cockroachdb/cockroach/issues/17514

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> after squash

---

Reviewed 2 of 15 files at r16, 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/server/updates.go, line 76 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kr1oPpbNsyPzTUBYAo-:-Kr1oPpbNsyPzTUBYAo0:b-nydpp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8057df569fd00918d166a80cc2666021c9f2d8e6/pkg/server/updates.go#L76)):*
> ```Go
> var (
> 
> // TODO(dt): this should be split from the report interval.
> ```

This comment will get lost here. We should move it with `diagnosticReportFrequency`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-:-Kr1ojTHfA8swvDJpTjI:bn98z5e)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented on pkg/settings/cluster/settings.go:
```diff
@@ -0,0 +1,656 @@
+}
+
+// MakeUpdater returns a new Updater, pre-alloced to the registry size.
+// Note that if the ClusterSetting has the Manual flag set, this Updater
```
if Settings

tschottdorf commented:
Review status: 143 of 145 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/server/updates.go, line 76 at r17](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486#-Kr1oPpbNsyPzTUBYAo-:-Kr1qY4DTK0W1FeddGkr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8057df569fd00918d166a80cc2666021c9f2d8e6/pkg/server/updates.go#L76)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This comment will get lost here. We should move it with `diagnosticReportFrequency`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17486)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented on pkg/settings/updater.go:
```diff
@@ -41,27 +41,50 @@ func EncodeFloat(f float64) string {
 }
 
-// Updater is a helper for updating the in-memory settings.
+// DefaultsUpdater is a helper for updating the in-memory settings.
```
DefaultsUpdater is a misnomer, we're not updating defaults..

tschottdorf replied:
But we are, the DefaultsUpdater resets everything that isn't explicitly set to its default value.

tschottdorf replied:
`ResettingUpdater`?

dt commented:
:lgtm:, with the understanding that the crash reporting and moving back to a empty base registry with runtime registration are both fast-follows, especially the former.

tschottdorf commented:
@RaduBerinde this just turned green and I'll prepare a follow-up PR anyway, I'll address your feedback there.

@dt understood, doing the former right now.

dt commented:
also, while I nitted those specifics, I neglected to give a huge 👍 on this overall -- I was regretting the singleton once I saw how it got used in practice and feel much better now. Thanks! 👏 👏 👏
